packages/orm/src/EntityManager.ts

```diff
@@ -482,30 +481,46 @@ export class EntityManager {
    * (i.e. including the initial `SELECT`s) in a transaction.
    */
   async flush(): Promise<void> {
-    if (this.currentFlushPromise) {
-      await this.currentFlushPromise;
-      await this.flushLoader.load(0);
-    } else {
-      await this.flushLoader.load(0);
+    if (this.isFlushing) {
+      throw new Error("Cannot flush while another flush is already in progress");
     }
-  }
 
-  /** The implementation of flush, but called by a DataLoader to de-dup calls made in a loop. */
-  private async doFlush(): Promise<void> {
-    // We defer doing this cascade logic until flush() so that delete() can remain synchronous.
-    await this.cascadeDeletesIntoUnloadedCollections();
-    recalcDerivedFields(this.entities);
-    const entityTodos = sortEntities(this.entities);
-    await addReactiveAsyncDerivedValues(entityTodos);
-    await recalcAsyncDerivedFields(this, entityTodos);
+    this._isFlushing = true;
+
+    // We need to split the thread so that Node's executionAsyncId changes and in turn
+    // Contexty give us a context that's distinct from whatever called flush
+    await new Promise((res) => setTimeout(res, 0));
+    const context = contexty.create();
+
+    const entitiesToFlush: Entity[] = [];
+    let pendingEntities = this.entities.filter((e) => e.isPendingFlush);
+
+    while (pendingEntities.length > 0) {
+      context.flushSecret = this.flushSecret;
+      // We defer doing this cascade logic until flush() so that delete() can remain synchronous.
+      await this.cascadeDeletesIntoUnloadedCollections(pendingEntities);
+      recalcDerivedFields(pendingEntities);
+      const entityTodos = sortEntities(pendingEntities);
+      await addReactiveAsyncDerivedValues(entityTodos);
+      await recalcAsyncDerivedFields(this, entityTodos);
+
+      // TODO Run beforeFlush first, so it can fill in derived values for validate?
+      await addReactiveValidations(entityTodos);
+      await validate(entityTodos);
+      await beforeDelete(entityTodos);
+      await beforeFlush(entityTodos);
+
+      entitiesToFlush.push(...pendingEntities);
```

Comment:
Per the `Set` comment, couldn't something in `pendingEntities` already be in `entitiesToFlush`? ...oh, I see, ont he next line, we explicitly don't add entities that were already in it. Cool.